### PR TITLE
De-neuter NODE_BLOOM

### PR DESCRIPTION
**Pull request by pstratem**
Disconnect nodes which request bloom filtering in violation of the protocol.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
See also https://github.com/bitcoin/bitcoin/pull/6579 for discussions.

utACK.

File: src/main.cpp [hunk 1]
```diff
@@ -4377,7 +4377,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         if (pfrom->nVersion >= NO_BLOOM_VERSION) {
             Misbehaving(pfrom->GetId(), 100);
             return false;
-        } else if (GetBoolArg("-enforcenodebloom", DEFAULT_ENFORCENODEBLOOM)) {
+        } else {
```
Comment (laanwj) [inline src/main.cpp]: DEFAULT_ENFORCENODEBLOOM constant in main.h needs to go too :)

**Comment by gmaxwell** [comment]
Has someone checked what impact this would have now? (Wouldn't preclude a merge in master, since there is still a fair amount of time to respond before release-- but it would be good to know).

**Comment by laanwj** [comment]
@gmaxwell Very little, I expect. the intersection of nodes that::

- Will immediately upgrade to code that includes this
- Set `-peerbloomfilters` false (it defaults to true, and will continue to do so)
- Are not yet using `-enforcenodebloom` on 0.12

Is likely to be very, very small.

This is just continuing the planning, in 0.12 NODE_BLOOM was introduced but not yet enforced (by default) for old versions, and the plan is that in 0.13 it will be enforced for all versions. 

What kind of measurement do you think should change that planning?

In no case this is going to be backported to 0.12.

**Comment by jonasschnelli** [comment]
Added a counter to my SPV stats node to see how many SPV nodes have not adapted the new rule.

**Comment by gmaxwell** [comment]
Ah. Indeed, I was forgetting that with peerbloomfilters defaulting to true this would have no effect. Still, interesting to know if there has been any progress or not.

**Comment by pstratem** [comment]
@laanwj nit picked

**Comment by laanwj** [comment]
ACK c90036f

**Comment by rebroad** [comment]
@pstratem What is the impact of not disconnecting these nodes?

**Comment by jonasschnelli** [comment]
@rebroad: check https://github.com/bitcoin/bips/blob/master/bip-0111.mediawiki.

Not all nodes want to provide the BLOOM FILTER service. SPV bloom filtering will consume lots of CPU and disk usage. The produced and consumed data will very likely be worthless for the p2p health after generating/transmitting (where a BLOCK or TX structure will be relay further by other full nodes to other full nodes).

BIP-0111 will help SPV clients to detect, which peers do and don't support the NODE_BLOOM service. Otherwise they need to try and disconnect in case they won't.

**Comment by rebroad** [comment]
@jonasnick @sipa thanks for explanaing

**Pull request merged**